File: übergabe/microexcel/microfilmC/project_service.c
/**
 * @file project_service.c
 * @brief Implementation of project management functions
 */

#include "microfilm.h"
#include <dirent.h>
#include <libgen.h>

static bool validate_path(const char *path, MicrofilmError *error);
static bool try_extract_metadata(const char *folder_name, char *archive_id, 
    char *location, char *doc_type);
static bool find_document_folder(const char *project_path, const char *archive_id, 
    char *doc_folder_path);
static bool find_comlist_file(const char *folder_path, const char *archive_id, 
    char *comlist_path);

MicrofilmError project_initialize(Project *project, const char *path, Logger *logger) {
    MicrofilmError error = {false, ""};
    
    if (!project || !path) {
        error.has_error = true;
        strcpy(error.message, "Invalid parameters");
        return error;
    }
    
    logger_section(logger, "Initializing Project");
    logger_log(logger, LOG_INFO, "PROJECT", "Processing path: %s", path);
    
    // Validate the path
    if (!validate_path(path, &error)) {
        return error;
    }
    
    // Copy the path and extract folder name
    strncpy(project->project_path, path, MAX_PATH_LENGTH - 1);
    
    // Extract folder name from path
    char *path_copy = strdup(path);
    char *folder_name = basename(path_copy);
    strncpy(project->project_folder_name, folder_name, MAX_PATH_LENGTH - 1);
    
    logger_log(logger, LOG_INFO, "PROJECT", "Folder name: %s", folder_name);
    
    // Try to extract metadata from folder name
    char archive_id[MAX_ARCHIVE_ID_LENGTH] = {0};
    char location[MAX_LOCATION_LENGTH] = {0};
    char doc_type[MAX_DOCTYPE_LENGTH] = {0};
    
    bool metadata_found = try_extract_metadata(folder_name, archive_id, location, doc_type);
    
    if (metadata_found) {
        // This is likely a project folder
        logger_log(logger, LOG_INFO, "PROJECT", "Path appears to be a project folder");
        
        strcpy(project->archive_id, archive_id);
        strcpy(project->location, location);
        strcpy(project->doc_type, doc_type);
        
        // Find document subfolder
        if (find_document_folder(path, archive_id, project->document_folder_path)) {
            char *doc_folder_name = basename(project->document_folder_path);
            strcpy(project->document_folder_name, doc_folder_name);
            logger_log(logger, LOG_INFO, "PROJECT", "Found document folder: %s", doc_folder_name);
        } else {
            logger_log(logger, LOG_WARNING, "PROJECT", "No document subfolder found");
        }
    } else {
        // Check if this might be a document subfolder
        char parent_path[MAX_PATH_LENGTH];
        strcpy(parent_path, path);
        char *parent_dir = dirname(parent_path);
        char *parent_name = basename(parent_dir);
        
        if (try_extract_metadata(parent_name, archive_id, location, doc_type)) {
            // This is a document subfolder
            logger_log(logger, LOG_INFO, "PROJECT", "Path appears to be a document subfolder");
            
            strcpy(project->archive_id, archive_id);
            strcpy(project->location, location);
            strcpy(project->doc_type, doc_type);
            strcpy(project->project_path, parent_dir);
            strcpy(project->project_folder_name, parent_name);
            strcpy(project->document_folder_path, path);
            strcpy(project->document_folder_name, folder_name);
        } else {
            error.has_error = true;
            snprintf(error.message, MAX_ERROR_MESSAGE_LENGTH,
                "Could not extract project metadata from folder name: %s or parent: %s",
                folder_name, parent_name);
            free(path_copy);
            return error;
        }
    }
    
    free(path_copy);
    
    // Find COM list file
    if (find_comlist_file(project->project_path, project->archive_id, project->comlist_path)) {
        logger_log(logger, LOG_INFO, "PROJECT", "Found COM list file: %s", 
            basename(project->comlist_path));
    }
    
    logger_log(logger, LOG_INFO, "PROJECT", "Project initialized: %s (%s)", 
        project->archive_id, project->location);
    
    return error;
}

static bool validate_path(const char *path, MicrofilmError *error) {
    if (!directory_exists(path)) {
        error->has_error = true;
        snprintf(error->message, MAX_ERROR_MESSAGE_LENGTH,
            "Path does not exist or is not a directory: %s", path);
        return false;
    }
    return true;
}

static bool try_extract_metadata(const char *folder_name, char *archive_id, 
    char *location, char *doc_type) {
    
    // Pattern: RRDxxx-xxxx_Location_DocType
    
    if (!folder_name || strlen(folder_name) < 12) {
        return false;
    }
    
    // Check if it starts with RRD
    if (strncmp(folder_name, "RRD", 3) != 0) {
        return false;
    }
    
    // Find first underscore
    const char *first_underscore = strchr(folder_name, '_');
    if (!first_underscore) {
        return false;
    }
    
    // Extract archive ID
    size_t archive_id_len = first_underscore - folder_name;
    if (archive_id_len >= MAX_ARCHIVE_ID_LENGTH) {
        return false;
    }
    strncpy(archive_id, folder_name, archive_id_len);
    archive_id[archive_id_len] = '\0';
    
    // Extract location
    const char *location_start = first_underscore + 1;
    const char *second_underscore = strchr(location_start, '_');
    
    size_t location_len;
    if (second_underscore) {
        location_len = second_underscore - location_start;
    } else {
        location_len = strlen(location_start);
    }
    
    if (location_len >= MAX_LOCATION_LENGTH) {
        return false;
    }
    strncpy(location, location_start, location_len);
    location[location_len] = '\0';
    
    // Extract doc type if available
    if (second_underscore) {
        const char *doc_type_start = second_underscore + 1;
        size_t doc_type_len = strlen(doc_type_start);
        if (doc_type_len < MAX_DOCTYPE_LENGTH) {
            strcpy(doc_type, doc_type_start);
        }
    }
    
    return true;
}

static bool find_document_folder(const char *project_path, const char *archive_id, 
    char *doc_folder_path) {
    
    DIR *dir = opendir(project_path);
    if (!dir) return false;
    
    struct dirent *entry;
    char full_path[MAX_PATH_LENGTH];
    bool found = false;
    
    // Strategy 1: Look for folder with "PDFs zu" in the name
    rewinddir(dir);
    while ((entry = readdir(dir)) != NULL) {
        if (entry->d_type == DT_DIR && strcmp(entry->d_name, ".") != 0 && 
            strcmp(entry->d_name, "..") != 0) {
            
            if (strstr(entry->d_name, "PDFs zu") != NULL) {
                snprintf(full_path, MAX_PATH_LENGTH, "%s/%s", project_path, entry->d_name);
                strcpy(doc_folder_path, full_path);
                found = true;
                break;
            }
        }
    }
    
    // Strategy 2: Look for folder containing the archive ID
    if (!found) {
        rewinddir(dir);
        while ((entry = readdir(dir)) != NULL) {
            if (entry->d_type == DT_DIR && strcmp(entry->d_name, ".") != 0 && 
                strcmp(entry->d_name, "..") != 0) {
                
                if (strstr(entry->d_name, archive_id) != NULL) {
                    snprintf(full_path, MAX_PATH_LENGTH, "%s/%s", project_path, entry->d_name);
                    strcpy(doc_folder_path, full_path);
                    found = true;
                    break;
                }
            }
        }
    }
    
    // Strategy 3: Look for folder containing PDF files
    if (!found) {
        rewinddir(dir);
        int max_pdfs = 0;
        char best_folder[MAX_PATH_LENGTH] = {0};
        
        while ((entry = readdir(dir)) != NULL) {
            if (entry->d_type == DT_DIR && strcmp(entry->d_name, ".") != 0 && 
                strcmp(entry->d_name, "..") != 0) {
                
                snprintf(full_path, MAX_PATH_LENGTH, "%s/%s", project_path, entry->d_name);
                
                // Count PDF files in this directory
                DIR *subdir = opendir(full_path);
                if (subdir) {
                    struct dirent *subentry;
                    int pdf_count = 0;
                    
                    while ((subentry = readdir(subdir)) != NULL) {
                        if (subentry->d_type == DT_REG) {
                            const char *ext = strrchr(subentry->d_name, '.');
                            if (ext && (strcasecmp(ext, ".pdf") == 0)) {
                                pdf_count++;
                            }
                        }
                    }
                    closedir(subdir);
                    
                    if (pdf_count > max_pdfs) {
                        max_pdfs = pdf_count;
                        strcpy(best_folder, full_path);
                    }
                }
            }
        }
        
        if (max_pdfs > 0) {
            strcpy(doc_folder_path, best_folder);
            found = true;
        }
    }
    
    closedir(dir);
    return found;
}

static bool find_comlist_file(const char *folder_path, const char *archive_id, 
    char *comlist_path) {
    
    DIR *dir = opendir(folder_path);
    if (!dir) return false;
    
    struct dirent *entry;
    char full_path[MAX_PATH_LENGTH];
    bool found = false;
    
    // Look for Excel files containing the archive ID
    while ((entry = readdir(dir)) != NULL) {
        if (entry->d_type == DT_REG) {
            const char *ext = strrchr(entry->d_name, '.');
            if (ext && (strcasecmp(ext, ".xls") == 0 || strcasecmp(ext, ".xlsx") == 0)) {
                // Check if filename contains archive ID
                if (strstr(entry->d_name, archive_id) != NULL) {
                    snprintf(full_path, MAX_PATH_LENGTH, "%s/%s", folder_path, entry->d_name);
                    strcpy(comlist_path, full_path);
                    found = true;
                    break;
                }
            }
        }
    }
    
    // If no specific match, look for any Excel file
    if (!found) {
        rewinddir(dir);
        while ((entry = readdir(dir)) != NULL) {
            if (entry->d_type == DT_REG) {
                const char *ext = strrchr(entry->d_name, '.');
                if (ext && (strcasecmp(ext, ".xls") == 0 || strcasecmp(ext, ".xlsx") == 0)) {
                    snprintf(full_path, MAX_PATH_LENGTH, "%s/%s", folder_path, entry->d_name);
                    strcpy(comlist_path, full_path);
                    found = true;
                    break;
                }
            }
        }
    }
    
    closedir(dir);
    return found;
}

File: übergabe/microexcel/microfilm/project_service.py
"""
Project service module for handling microfilm project operations.

This module provides a service layer for project-related operations, 
including initialization, detection, and management. It's designed to be 
reusable across different interfaces (CLI, web, etc.).
"""

import os
import re
import json
import logging
from pathlib import Path
from typing import Dict, List, Optional, Tuple, Any, Union

from models import Project
from exceptions import (
    ProjectError, ProjectInitializationError, ProjectLoadError, 
    ProjectExportError, DirectoryError, FileError, ValidationError
)


class ProjectService:
    """
    Service for handling all project-related operations.
    
    This service abstracts all logic related to project initialization,
    detection, management, and serialization, making it reusable across
    different interfaces like CLI, web applications, etc.
    """
    
    def __init__(self, logger=None):
        """
        Initialize the project service.
        
        Args:
            logger: Optional logger instance. If None, logging will be disabled.
                   Can be a standard logging.Logger or a FilmLogger instance.
        """
        self.logger = logger
    
    def initialize_project(self, path: Union[str, Path]) -> Project:
        """
        Initialize a project from a file system path.
        
        This method can handle both scenarios:
        1. User selects the project folder directly
        2. User selects the document subfolder
        
        Args:
            path: Path to either the project folder or document subfolder
            
        Returns:
            Initialized Project object
            
        Raises:
            ValidationError: If the project cannot be initialized
        """
        self.logger.section("Initializing Project")
        try:
            # Convert to Path object if string
            if isinstance(path, str):
                path = Path(path)
            
            # Validate the path
            if not path.exists():
                raise ValidationError(f"Path does not exist: {path}")
            if not path.is_dir():
                raise ValidationError(f"Path is not a directory: {path}")
            
            # Determine if path is project folder or document subfolder
            folder_name = path.name
            self.logger.project_info(f"Processing path: {path}, folder name: {folder_name}")
            
            # Try to extract metadata from the folder name
            project_metadata = self._try_extract_metadata(folder_name)
            
            if project_metadata:
                # This is likely a project folder
                self.logger.project_info(f"Path appears to be a project folder")
                archive_id, location, doc_type = project_metadata
                
                # Create project with this as project folder
                project = Project(
                    archive_id=archive_id,
                    location=location,
                    project_path=path,
                    project_folder_name=folder_name,
                    doc_type=doc_type
                )
                
                # Find document subfolder
                doc_folder = self._find_document_folder(path, archive_id)
                if doc_folder:
                    project.document_folder_path = doc_folder
                    project.document_folder_name = doc_folder.name
                    self.logger.project_info(f"Found document folder: {doc_folder.name}")
                else:
                    self.logger.project_warning(f"No document subfolder found in project folder")
            else:
                # This might be a document subfolder, check parent
                parent_path = path.parent
                parent_name = parent_path.name
                
                parent_metadata = self._try_extract_metadata(parent_name)
                if parent_metadata:
                    # This is a document subfolder
                    self.logger.project_info(f"Path appears to be a document subfolder")
                    archive_id, location, doc_type = parent_metadata
                    
                    # Create project
                    project = Project(
                        archive_id=archive_id,
                        location=location,
                        project_path=parent_path,
                        project_folder_name=parent_name,
                        document_folder_path=path,
                        document_folder_name=folder_name,
                        doc_type=doc_type
                    )
                else:
                    # Neither folder has metadata, can't initialize
                    raise ValidationError(
                        f"Could not extract project metadata from folder name: {folder_name} "
                        f"or parent folder: {parent_name}"
                    )
            
            # Find COM list file
            comlist_path = self._find_comlist_file(project.project_path, project.archive_id)
            if comlist_path:
                project.comlist_path = comlist_path
                self.logger.project_info(f"Found COM list file: {comlist_path.name}")
            
            self.logger.project_info(f"Project initialized: {project.archive_id} ({project.location})")
            return project
            
        except ValidationError as e:
            self.logger.project_error(f"Validation error during project initialization: {str(e)}")
            raise
        except Exception as e:
            error_msg = f"Error initializing project: {str(e)}"
            self.logger.project_error(error_msg)
            raise ProjectInitializationError(error_msg)
    
    def _try_extract_metadata(self, folder_name: str) -> Optional[Tuple[str, str, str]]:
        """
        Try to extract project metadata from a folder name.
        
        Args:
            folder_name: Name of the folder to extract metadata from
            
        Returns:
            Tuple of (archive_id, location, doc_type) if successful, None otherwise
        """
        try:
            return self._extract_metadata_from_folder_name(folder_name)
        except ValidationError:
            return None
    
    def _extract_metadata_from_folder_name(self, folder_name: str) -> Tuple[str, str, str]:
        """
        Extract archive ID, location, and document type from folder name.
        
        Expected format: RRDxxx-xxxx_Location_DocType
        Example: RRD018-2024_OU_FAIR
        
        Args:
            folder_name: Name of the folder to extract metadata from
            
        Returns:
            Tuple of (archive_id, location, doc_type)
            
        Raises:
            ValidationError: If the folder name doesn't match the expected format
        """
        # Regular expression to match the expected format
        pattern = r'^(RRD\d{3}-\d{4})_([A-Z]{2})_(.+)$'
        match = re.match(pattern, folder_name)
        
        if not match:
            # Try alternative pattern without document type
            alt_pattern = r'^(RRD\d{3}-\d{4})_([A-Z]{2})$'
            alt_match = re.match(alt_pattern, folder_name)
            
            if alt_match:
                archive_id, location = alt_match.groups()
                doc_type = ""
            else:
                raise ValidationError(f"Invalid folder name format: {folder_name}. "
                                f"Expected format: RRDxxx-xxxx_Location_DocType")
        else:
            archive_id, location, doc_type = match.groups()
        
        # Validate location
        if location not in ['OU', 'DW'] and self.logger:
            self.logger.project_warning(f"Unexpected location code: {location}. Expected 'OU' or 'DW'.")

        return archive_id, location, doc_type
    
    def _find_document_folder(self, project_path: Path, archive_id: str) -> Optional[Path]:
        """
        Find the document subfolder using enhanced two-layer detection strategy.
        
        Layer 1: String detection (pdf|archive_id -> importiert -> any folder)
        Layer 2: File validation for confirmation and conflict resolution
        
        Args:
            project_path: Path to the project folder
            archive_id: Archive ID like "RRD018-2024"
            
        Returns:
            Path to the document subfolder if found, None otherwise
        """
        if self.logger:
            self.logger.project_info("Starting enhanced document folder detection")
        
        # Get all subdirectories (excluding hidden folders)
        subdirs = [item for item in project_path.iterdir() 
                  if item.is_dir() and not item.name.startswith('.')]
        
        if not subdirs:
            if self.logger:
                self.logger.project_warning("No subdirectories found in project folder")
            return None
        
        # === LAYER 1: STRING DETECTION ===
        
        # First priority: PDF patterns or Archive ID
        pdf_or_archive_candidates = []
        for folder in subdirs:
            folder_name = folder.name.lower()
            
            # Check for PDF patterns (case-insensitive)
            if re.search(r'pdf[s]?[\s_-]*(zu|zum|für|verfilmung)?', folder_name):
                pdf_or_archive_candidates.append(folder)
            # Check for archive ID
            elif archive_id.lower() in folder_name:
                pdf_or_archive_candidates.append(folder)
        
        # Second priority: importiert pattern
        importiert_candidates = []
        for folder in subdirs:
            if re.search(r'importiert', folder.name, re.IGNORECASE):
                importiert_candidates.append(folder)
        
        # Final fallback: any remaining folders
        other_candidates = [f for f in subdirs 
                          if f not in pdf_or_archive_candidates + importiert_candidates]
        
        if self.logger:
            self.logger.project_info(f"String detection found: "
                                   f"{len(pdf_or_archive_candidates)} PDF/Archive, "
                                   f"{len(importiert_candidates)} importiert, "
                                   f"{len(other_candidates)} other folders")
        
        # === LAYER 2: FILE VALIDATION ===
        
        def analyze_folder_contents(folder_path: Path) -> dict:
            """Analyze folder contents and return file statistics"""
            try:
                all_files = [f for f in folder_path.iterdir() if f.is_file()]
                pdf_files = [f for f in all_files if f.suffix.lower() == '.pdf']
                
                return {
                    'path': folder_path,
                    'total_files': len(all_files),
                    'pdf_files': len(pdf_files),
                    'only_pdfs': len(pdf_files) == len(all_files) and len(all_files) > 0,
                    'has_pdfs': len(pdf_files) > 0,
                    'non_pdf_files': len(all_files) - len(pdf_files),
                    'is_empty': len(all_files) == 0
                }
            except (PermissionError, OSError) as e:
                if self.logger:
                    self.logger.project_warning(f"Cannot access folder {folder_path}: {e}")
                return {
                    'path': folder_path,
                    'total_files': 0,
                    'pdf_files': 0,
                    'only_pdfs': False,
                    'has_pdfs': False,
                    'non_pdf_files': 0,
                    'is_empty': True,
                    'access_error': True
                }
        
        # Analyze all candidates
        all_candidates = pdf_or_archive_candidates + importiert_candidates + other_candidates
        folder_analysis = {folder: analyze_folder_contents(folder) for folder in all_candidates}
        
        # === DECISION LOGIC ===
        
        # First priority: PDF/Archive candidates
        if pdf_or_archive_candidates:
            valid_pdf_archive = []
            
            for folder in pdf_or_archive_candidates:
                analysis = folder_analysis[folder]
                
                # Skip empty folders (FAIL condition)
                if analysis['is_empty']:
                    if self.logger:
                        self.logger.project_warning(f"Skipping empty PDF/Archive folder: {folder.name}")
                    continue
                
                # OK conditions:
                # - Only PDFs
                # - Has PDFs + other files (but non-PDF files < 5)
                if analysis['only_pdfs'] or (analysis['has_pdfs'] and analysis['non_pdf_files'] < 5):
                    valid_pdf_archive.append((folder, analysis))
            
            if valid_pdf_archive:
                # Conflict resolution: prefer folders with only PDFs
                only_pdf_folders = [(f, a) for f, a in valid_pdf_archive if a['only_pdfs']]
                
                if only_pdf_folders:
                    # Return the first folder with only PDFs
                    chosen_folder = only_pdf_folders[0][0]
                    if self.logger:
                        self.logger.project_success(f"Selected PDF/Archive folder with only PDFs: {chosen_folder.name}")
                    return chosen_folder
                else:
                    # Return the first valid folder (has PDFs + few other files)
                    chosen_folder = valid_pdf_archive[0][0]
                    analysis = valid_pdf_archive[0][1]
                    if self.logger:
                        self.logger.project_success(f"Selected PDF/Archive folder: {chosen_folder.name} "
                                                   f"({analysis['pdf_files']} PDFs, {analysis['non_pdf_files']} other files)")
                    return chosen_folder
        
        # Second priority: importiert candidates (only if no valid PDF/Archive found)
        if importiert_candidates:
            for folder in importiert_candidates:
                analysis = folder_analysis[folder]
                
                # Skip empty folders (FAIL condition)
                if analysis['is_empty']:
                    if self.logger:
                        self.logger.project_warning(f"Skipping empty importiert folder: {folder.name}")
                    continue
                
                # OK if has PDFs (same rules as PDF/Archive)
                if analysis['only_pdfs'] or (analysis['has_pdfs'] and analysis['non_pdf_files'] < 5):
                    if self.logger:
                        self.logger.project_success(f"Selected importiert folder: {folder.name} "
                                                   f"({analysis['pdf_files']} PDFs, {analysis['non_pdf_files']} other files)")
                    return folder
        
        # Final fallback: any folder with only PDFs (if no other valid options)
        if other_candidates:
            only_pdf_others = []
            
            for folder in other_candidates:
                analysis = folder_analysis[folder]
                
                if analysis['only_pdfs']:
                    only_pdf_others.append((folder, analysis))
            
            # OK if exactly one folder with only PDFs exists
            if len(only_pdf_others) == 1:
                chosen_folder = only_pdf_others[0][0]
                if self.logger:
                    self.logger.project_success(f"Selected fallback folder with only PDFs: {chosen_folder.name}")
                return chosen_folder
        
        # No valid document folder found
        if self.logger:
            self.logger.project_error("No valid document folder found using enhanced detection")
            
            # Log detailed analysis for debugging
            for folder, analysis in folder_analysis.items():
                status = "EMPTY" if analysis['is_empty'] else f"{analysis['pdf_files']}PDF/{analysis['non_pdf_files']}OTHER"
                self.logger.project_info(f"  {folder.name}: {status}")
        
        return None
    
    def _find_comlist_file(self, folder_path: Path, archive_id: str) -> Optional[Path]:
        """
        Find the COM list Excel file in the given folder using enhanced detection.
        
        Requirements:
        1. Must contain both "com" and "list" keywords (case-insensitive)
        2. Nice to have: contains the archive ID (RRD pattern)
        3. Must be Excel file (.xls, .xlsx, .xlsb, .xlsm)
        
        Args:
            folder_path: Path to the folder to search in
            archive_id: Archive ID to look for in filenames
            
        Returns:
            Path to the COM list file if found, None otherwise
        """
        try:
            # Look for Excel files in the folder (strict extensions)
            excel_extensions = ['*.xls', '*.xlsx', '*.xlsb', '*.xlsm']
            excel_files = []
            for ext in excel_extensions:
                excel_files.extend(folder_path.glob(ext))
            
            if not excel_files:
                if self.logger:
                    self.logger.project_warning(f"No Excel files found in {folder_path}")
                return None
            
            if self.logger:
                self.logger.project_info(f"Found {len(excel_files)} Excel files to analyze")
            
            # === ENHANCED COM LIST DETECTION ===
            # Priority order: "com" > "list" > archive_id
            # Accept files containing ANY of these keywords
            
            # Step 1: Categorize files by keyword priority
            com_files = []
            list_files = []
            archive_files = []
            
            archive_id_pattern = re.compile(re.escape(archive_id), re.IGNORECASE)
            
            for excel_file in excel_files:
                filename_lower = excel_file.name.lower()
                
                # Check for keywords in priority order
                if 'com' in filename_lower:
                    com_files.append(excel_file)
                    if self.logger:
                        self.logger.project_info(f"COM keyword found: {excel_file.name}")
                elif 'list' in filename_lower:
                    list_files.append(excel_file)
                    if self.logger:
                        self.logger.project_info(f"LIST keyword found: {excel_file.name}")
                elif archive_id_pattern.search(excel_file.name):
                    archive_files.append(excel_file)
                    if self.logger:
                        self.logger.project_info(f"Archive ID found: {excel_file.name}")
            
            # Step 2: Select best match based on priority
            if com_files:
                selected_file = com_files[0]
                if self.logger:
                    self.logger.project_success(f"Found COM list file (priority: COM): {selected_file.name}")
                    if len(com_files) > 1:
                        self.logger.project_info(f"Multiple COM files found, using first: {selected_file.name}")
                return selected_file
            
            elif list_files:
                selected_file = list_files[0]
                if self.logger:
                    self.logger.project_success(f"Found COM list file (priority: LIST): {selected_file.name}")
                    if len(list_files) > 1:
                        self.logger.project_info(f"Multiple LIST files found, using first: {selected_file.name}")
                return selected_file
            
            elif archive_files:
                selected_file = archive_files[0]
                if self.logger:
                    self.logger.project_success(f"Found COM list file (priority: ARCHIVE_ID): {selected_file.name}")
                    if len(archive_files) > 1:
                        self.logger.project_info(f"Multiple archive ID files found, using first: {selected_file.name}")
                return selected_file
            
            else:
                # No matching files found
                if self.logger:
                    self.logger.project_error(f"No Excel files found containing 'com', 'list', or archive ID '{archive_id}' in project {archive_id}")
                    self.logger.project_info("Available Excel files:")
                    for f in excel_files:
                        self.logger.project_info(f"  - {f.name}")
                return None
            
        except Exception as e:
            if self.logger:
                self.logger.project_error(f"Error finding COM list file: {str(e)}")
            return None
        
    def configure_logger(self, project: Project, logger):
        """
        Configure the logger to save logs in the project directory.
        """
        if project.project_path:
            project_log_dir = os.path.join(str(project.project_path), '.logs')
            logger.parent_folder = project_log_dir
            logger.project_info(f"Logs will be saved to: {project_log_dir}")
            logger.save_log_file(archive_id=project.archive_id)
